### PR TITLE
♻️ Configuration assembly refactor

### PR DESCRIPTION
**Pull request by muchzill4**
### Description

I've realised that layer and modifier concepts were an unnecessary abstraction - there was a lot of data processing for no good reason. I believe that this might be a more straightforward approach. Granted all the config merging happens in a loop, but at least it's in one place, where it should be easy to debug (something I struggled with a bit with the previous implementation).

What makes this slightly convoluted is the fact that target data is pre-processed by `mbed-targets` (which I'm grateful for!) and I'm squeezing it into a `Source` shape.

If you've reviewed the previous code it should be familiar, if not:

`Config` is obviously what we're building here. It has:
- `Option` s - actual configuration options
- `TargetMetadata` these are the infamous cumulative overrides which affect filtering rules when discovering `mbed_lib.json` files

`Source` is an abstraction over _configuration source_, be it `mbed_targets.Target` or `mbed_lib.json` and `mbed_app.json` files (to be implemented).

_Big_ changes from previous way of doing things:
- Tightened the config assembly functionality - you should not be able to apply a "target_override" over a config entry which does not exist
- Added config option tracing - we now know where given option comes from - could be useful for config per lib approach, is now useful for seeing where the hell given `Option` came from

FWIW I've been testing it with single pass config assembly, but decided to drop that from this PR for the sake of keeping the diff smaller (it's already inflated). You can check it out here: https://github.com/ARMmbed/mbed-build/pull/27/commits/152cf1c30000e7d448e4a2b618a10ab877413380

### Test Coverage

- [x]  This change is covered by existing or additional automated tests.
- [ ]  Manual testing has been performed (and evidence provided) as automated testing was not feasible.
- [ ]  Additional tests are not required for this change (e.g. documentation update).

### CODE REVIEW

File: mbed_build/_internal/config/assemble_config.py [hunk 1]
```diff
@@ -0,0 +1,17 @@
...
+from mbed_build._internal.config.config import Config
+
+
+def assemble_config(mbed_target: str, mbed_program_directory: pathlib.Path) -> Config:
+    """Assemble Config for given target and program directory.
+
+    The structure and configuration of MbedOS requires us to do multiple passes over
+    configuration assembly, as each Source can modify filtering rules applied to finding libs.
+    """
+    return Config()
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/assemble_config.py:17]: Is this WIP? I suppose you have another PR ready which passes the arguments through?

File: mbed_build/_internal/config/config.py [hunk 2]
```diff
@@ -2,47 +2,122 @@
...
+                set_by=source.name,
+            )
+        else:
+            return cls(value=data, macro_name=None, help_text=None, set_by=source.name)
+
+    def set_value(self, value: Any, source: Source) -> None:
+        """Mutate self with new value."""
+        self.value = value
+        self.set_by = source.name
+        return self
```
Comment (rwalton-arm) [inline mbed_build/_internal/config/config.py:44]: Looks like a typo here?

File: mbed_build/_internal/config/assemble_config.py (hunk 1 above)
Comment (muchzill4) [reply mbed_build/_internal/config/assemble_config.py:17]: Yeah, I had this in first, but decided to drop the code as it was untested. Planning to add this as next step. Do you think this file should be removed? 🤔

Comment (muchzill4) [reply mbed_build/_internal/config/assemble_config.py:17]: FWIW, rationale was - this allows for a separate thread where a command to display config can be implemented. `assemble_config` is the interface to retrieve it.

File: mbed_build/_internal/config/config.py (hunk 2 above)
Comment (muchzill4) [reply mbed_build/_internal/config/config.py:44]: ? 🤔

Comment (rwalton-arm) [reply mbed_build/_internal/config/config.py:44]: We shouldn't need to return `self` here, and the type hint says we return `None`.

Comment (muchzill4) [reply mbed_build/_internal/config/config.py:44]: Ah, you're right. `self` is there for the sake of using this like a fluent interface. I'll fix the type hint.

Comment (muchzill4) [reply mbed_build/_internal/config/config.py:44]: The bigger question is - how the heck does ci not fail?

Comment (muchzill4) [reply mbed_build/_internal/config/config.py:44]: Answer: missing `__init__.py` file. https://github.com/ARMmbed/mbed-build/pull/27/commits/bf7c4bcc620e44ae69df62d46826d6d0cfc1d624

**Pull request merged**